# Importing Data

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\iNDIA\Desktop\Career Foundry\4 Python Fundamentals for Data Analysts\Instacart Basket Analysis Mar 22\02 Data\Original Data\products.csv'

In [3]:
# Importing Products from Original Data
df_prods = pd.read_csv(os.path.join(path), index_col = False)

In [4]:
orderspath=r'C:\Users\iNDIA\Desktop\Career Foundry\4 Python Fundamentals for Data Analysts\Instacart Basket Analysis Mar 22\02 Data\Prepared Data\orders_wrangled.csv'

In [5]:
#Importing Orders
df_ords = pd.read_csv(os.path.join(orderspath), index_col = False)

In [6]:
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [7]:
# Change types for ords data set

df_prods['product_id']=df_prods['product_id'].astype('int32')
df_prods['aisle_id'] = df_prods['aisle_id'].astype('int32')
df_prods['department_id']=df_prods['department_id'].astype('int16')
df_prods['prices']=df_prods['prices'].astype('float64')

In [8]:
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int32  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int32  
 3   department_id  49693 non-null  int16  
 4   prices         49693 non-null  float64
dtypes: float64(1), int16(1), int32(2), object(1)
memory usage: 1.2+ MB


In [9]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_day_of_week       int64  
 5   order_time              int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [10]:
df_ords['order_id'] = df_ords['order_id'].astype('int32')
df_ords['user_id'] = df_ords['user_id'].astype('int16')
df_ords['order_number'] = df_ords['order_number'].astype('int16')
df_ords['order_day_of_week'] = df_ords['order_day_of_week'].astype('int16')
df_ords['order_time'] = df_ords['order_time'].astype('int16')
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].astype('float16')

In [11]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int32  
 2   user_id                 int16  
 3   order_number            int16  
 4   order_day_of_week       int16  
 5   order_time              int16  
 6   days_since_prior_order  float16
dtypes: float16(1), int16(4), int32(1), int64(1)
memory usage: 71.8 MB


# Data Consistency Checks

In [12]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  2.604052e+02  1.715486e+01   
std    9.875817e+05  9.875817e+05  1.849610e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00 -3.276800e+04  1.000000e+00   
25%    8.552705e+05  8.552715e+05 -1.553700e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.261000e+03  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.561600e+04  2.300000e+01   
max    3.421082e+06  3.421083e+06  3.276700e+04  1.000000e+02   

       order_day_of_week    order_time  days_since_prior_order  
count       3.421083e+06  3.421083e+06               3214874.0  
mean        2.776219e+00  1.345202e+01                     NaN  
std         2.046829e+00  4.226088e+00                     0.0  
min         0.000000e+00  0.000000e+00                     0.0  
25%         1.000000e+00  1.000000e+01                     4.0  
50%         3.000000e+00  1.300000e+01                     7.0  
75%         5.000000e+00  1.600000e+01                    15.0  
max         6.000000e+00  2.300000e+01                    30.0

In [13]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [14]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_time  \
0   2539329        1             1                  2           8   
1   2398795        1             2                  3           7   
2    473747        1             3                  3          12   
3   2254736        1             4                  4           7   
4    431534        1             5                  4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [15]:
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  2.604052e+02  1.715486e+01       2.776219e+00   
std    9.875817e+05  1.849610e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00 -3.276800e+04  1.000000e+00       0.000000e+00   
25%    8.552715e+05 -1.553700e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.261000e+03  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.561600e+04  2.300000e+01       5.000000e+00   
max    3.421083e+06  3.276700e+04  1.000000e+02       6.000000e+00   

         order_time  days_since_prior_order  
count  3.421083e+06               3214874.0  
mean   1.345202e+01                     NaN  
std    4.226088e+00                     0.0  
min    0.000000e+00                     0.0  
25%    1.000000e+01                     4.0  
50%    1.300000e+01                     7.0  
75%    1.600000e+01                    15.0  
max    2.300000e+01                    30.0

In [16]:
# Create a dataframe
df_test = pd.DataFrame()

In [17]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [18]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [19]:
# Check for mixed type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [20]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [21]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
# View missing dataset
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [23]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [24]:
df_prods.median()

C:\Users\iNDIA\AppData\Local\Temp/ipykernel_6912/3387230063.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [25]:
df_prods.shape

(49693, 5)

In [26]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
df_prods_2 = df_prods.dropna(subset = ['product_name'],inplace = True)

# Duplicates

In [29]:
# looking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [30]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [31]:
df_prods_clean.shape

(49677, 5)

In [32]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

In [34]:
path_products = r'C:\Users\iNDIA\Desktop\Career Foundry\4 Python Fundamentals for Data Analysts\Instacart Basket Analysis Mar 22\02 Data\Prepared Data'

In [35]:
df_prods.to_csv(os.path.join(path_products, 'products_checked.csv'))

# TASK

In [36]:
# Q2. Run the df.describe() function on your df_prods dataframe.

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

A price of 99999 looks much higher for a product. We will need to check it with the client

In [37]:
df_prods['prices'].max()

99999.0

In [38]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [39]:
df_prods.loc[df_prods['prices'] > 100, 'prices'] =np.nan

In [40]:
df_prods['prices'].max()

25.0

In [41]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49675.000000
mean   24850.194235     67.763110      11.728687      7.680552
std    14340.588602     38.316396       5.850651      4.199287
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [42]:
# Q3. Check for mixed-type data in your df_ords dataframe.

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc
    [0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

In [43]:
# Q4. If you find mixed-type data, fix it. 
# The column in question should contain observations of a single data type.

##No mix data found in df_ords dataframe

In [44]:
# Q5. Run a check for missing values in your df_ords dataframe

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_time                     0
days_since_prior_order    206209
dtype: int64

days_since_prior_order has the missing values (206209), this might be due to the customers data not updated or most of the customers being first time customers

In [45]:
# Q6. Address the missing values using an appropriate method.

df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [46]:
df_nan

order_id  user_id  order_number  order_day_of_week  order_time  \
0         2539329        1             1                  2           8   
11        2168274        2             1                  2          11   
26        1374495        3             1                  1          14   
39        3343014        4             1                  6          11   
45        2717275        5             1                  3          12   
...           ...      ...           ...                ...         ...   
3420930    969311     9597             1                  4          12   
3420934   3189322     9598             1                  3          18   
3421002   2166133     9599             1                  6          19   
3421019   2227043     9600             1                  1          15   
3421069   3154581     9601             1                  3          11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

We won't be deleting this data as this data can be updated as soon as the order is made by that partical user (user_id), but we can add another column (dataframe) to highlight the columns with missing values as first_time customers

In [47]:
# creating new dataframe for cleaned data 

df_ords_clean= df_ords

In [48]:
df_ords_clean['first_time_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [49]:
df_ords_clean.head()

order_id  user_id  order_number  order_day_of_week  order_time  \
0   2539329        1             1                  2           8   
1   2398795        1             2                  3           7   
2    473747        1             3                  3          12   
3   2254736        1             4                  4           7   
4    431534        1             5                  4          15   

   days_since_prior_order  first_time_customer  
0                     NaN                 True  
1                    15.0                False  
2                    21.0                False  
3                    29.0                False  
4                    28.0                False

In [50]:
# Q7. Run a check for duplicate values in your df_ords data

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [51]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_time, days_since_prior_order, first_time_customer]
Index: []

no duplicate found in the database

In [52]:
# Q9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

path_order_checked = r'C:\Users\iNDIA\Desktop\Career Foundry\4 Python Fundamentals for Data Analysts\Instacart Basket Analysis Mar 22\02 Data\Prepared Data'

In [ ]:
%%time
df_ords_clean.to_csv(os.path.join(path_order_checked, 'orders_checked.csv'))

In [ ]:
%%time
df_ords_clean.isnull().sum()